pkg/cmd/pr/merge/merge.go

```diff
@@ -168,224 +168,326 @@ func NewCmdMerge(f *cmdutil.Factory, runF func(*MergeOptions) error) *cobra.Comm
 	}
 
-	branchSwitchString := ""
-
-	if opts.CanDeleteLocalBranch && localBranchExists {
-		currentBranch, err := opts.Branch()
+	if m.opts.CanDeleteLocalBranch && m.localBranchExists {
+		currentBranch, err := m.opts.Branch()
 		if err != nil {
 			return err
 		}
 
-		var branchToSwitchTo string
-		if currentBranch == pr.HeadRefName {
-			branchToSwitchTo = pr.BaseRefName
-			if branchToSwitchTo == "" {
-				branchToSwitchTo, err = api.RepoDefaultBranch(apiClient, baseRepo)
+		// branch the command was run on is the same as the pull request branch
+		if currentBranch == m.pr.HeadRefName {
+			// if the target branch of the PR is not known, set the current branch to the
+			// default branch of the repository
+			m.currentBranch = m.pr.BaseRefName
+			if m.currentBranch == "" {
+				m.currentBranch, err = api.RepoDefaultBranch(m.apiClient, m.baseRepo)
 				if err != nil {
 					return err
 				}
 			}
 
-			remotes, err := opts.Remotes()
+			remotes, err := m.opts.Remotes()
 			if err != nil {
 				return err
 			}
 
-			baseRemote, err := remotes.FindByRepo(baseRepo.RepoOwner(), baseRepo.RepoName())
+			baseRemote, err := remotes.FindByRepo(m.baseRepo.RepoOwner(), m.baseRepo.RepoName())
 			if err != nil {
 				return err
 			}
 
-			if git.HasLocalBranch(branchToSwitchTo) {
-				if err := git.CheckoutBranch(branchToSwitchTo); err != nil {
+			if git.HasLocalBranch(m.currentBranch) {
+				if err := git.CheckoutBranch(m.currentBranch); err != nil {
 					return err
 				}
 			} else {
-				if err := git.CheckoutNewBranch(baseRemote.Name, branchToSwitchTo); err != nil {
+				if err := git.CheckoutNewBranch(baseRemote.Name, m.currentBranch); err != nil {
 					return err
 				}
 			}
 
-			if err := git.Pull(baseRemote.Name, branchToSwitchTo); err != nil {
-				fmt.Fprintf(opts.IO.ErrOut, "%s warning: not possible to fast-forward to: %q\n", cs.WarningIcon(), branchToSwitchTo)
+			if err := git.Pull(baseRemote.Name, m.currentBranch); err != nil {
+				m.log(fmt.Sprintf("%s warning: not possible to fast-forward to: %q\n", m.cs.WarningIcon(), m.currentBranch))
 			}
 		}
 
-		if err := git.DeleteLocalBranch(pr.HeadRefName); err != nil {
-			err = fmt.Errorf("failed to delete local branch %s: %w", cs.Cyan(pr.HeadRefName), err)
+		if err := git.DeleteLocalBranch(m.pr.HeadRefName); err != nil {
+			err = fmt.Errorf("failed to delete local branch %s: %w", m.cs.Cyan(m.pr.HeadRefName), err)
 			return err
 		}
+	}
 
-		if branchToSwitchTo != "" {
-			branchSwitchString = fmt.Sprintf(" and switched to branch %s", cs.Cyan(branchToSwitchTo))
-		}
+	return nil
+}
+
+// Delete the remote branch if requested and if allowed.
+func (m *MergeContext) deleteRemoteBranch() error {
+	// the user was already asked if they want to delete the branch if they didn't provide the flag
+	if !m.deleteBranch || m.crossRepoPR || m.autoMerge {
+		return nil
 	}
-	if !isPRAlreadyMerged {
-		err = api.BranchDeleteRemote(apiClient, baseRepo, pr.HeadRefName)
+
+	if !m.merged {
```

Comment:
I'm confused by this condition ([prior to refactor](https://github.com/cli/cli/blob/76782744648658e1eee9ee8a666f01d554ac9a16/pkg/cmd/pr/merge/merge.go#L377))

I think we should only be trying to delete the remote branch after the PR is merged, but the condition checks the opposite